### DISCUSSION TITLE
Detect inefficient math mixing float and int

### DISCUSSION THREAD

**Discussion by cspotcode** [microsoft/pyright#4606] [Ideas]
Is it feasible for pyright to detect when code does something like `my_float * 2` and suggest changing `2` to `2.0` to take advantage of python 3.11's ["Specializing Adapting Interpreter"](https://docs.python.org/3/whatsnew/3.11.html#pep-659-specializing-adaptive-interpreter)?

Multiplying a float by an int isn't a type error, so pyright shouldn't really flag this; it's out of scope.  Nevertheless, for a linter to flag this kind of thing, it needs to have typechecking machinery, so I feel like this can only be accomplished by pyright or a pyright plugin.

I'm curious if anyone else has thoughts on this.  Maybe a fork of pyright is the best approach, re-using as much pyright as possible, adding a diagnostic check for when `int` and `float` types are mixed in math operators.

**Comment by erictraut** [comment]
I consider this out of scope for a static type checker. This is more of a linter feature, so you might want to reach out to the maintainers of pylint or other popular Python linters with this request.

**Reply by cspotcode** [reply]
That makes sense and I agree.  Do you know if any linters do their own type analysis sufficient to know the static types of both sides of a binary operator?  I feel like probably only pyright and mypy have the necessary analysis to know.

**Reply by r3m0t** [reply]
Hi cspotcode, pylint/astroid has inference and a good tutorial on writing checkers. However it predates type annotations so I'm not sure whether it respects these. https://pylint.pycqa.org/en/latest/development_guide/contributor_guide/index.html /  https://pylint.pycqa.org/projects/astroid/en/latest/inference.html
